### ISSUE TITLE
zsh completion/fpath not set in mosh/tmux

### ISSUE THREAD

**Issue by Mic92** [NixOS/nixpkgs#80437]
**Describe the bug**

When using mosh/tmux `/etc/zshenv` is first read by the login shell
that spawns mosh-server, which sets `__ETC_ZSHENV_SOURCED=1`.
The actual user shell spawned by mosh will therefor ignore this file and not set its fpath. For that reason all system provided completion functions are ignored.

**Comment by Mic92** [comment]
This bug was introduced by @ttuegel in 6a756af3e7a50c22a45f7d958657527ac47d2902
who changed this variable to `export`

**Comment by ttuegel** [comment]
> What was the reasoning for not changing the environment of interactive non-login shells?

Um, this was six years ago, but I will attempt to answer the question. :sweat_smile: Consider this scenario:

1. You log in and the shell environment is initialized.
1. You make some local changes to the environment, and export them.
1. You open a sub-shell and the shell environment is re-initialized.
1. The changes made in the parent shell environment are not present in the sub-shell, as one would expect.

Basically, the zsh initialization process would clobber any environment that you tried to propagate from a parent shell into a child sub-shell.

**Comment by Mic92** [comment]
Currently zshenv by default only set fpath and HELPDIR without exporting them. 
A parent shell would also not set those variables usually as they are shell local.

It also sources a file called set-environment but this is protected by an environment variable called `__NIXOS_SET_ENVIRONMENT_DONE`. Hence any modification done by the parent shell should persist as long as __NIXOS_SET_ENVIRONMENT_DONE is not unset.

This behavior deviates from what we do in bashrc. Hence I would revert it.

**Comment by ttuegel** [comment]
> Hence any modification done by the parent shell should persist as long as __NIXOS_SET_ENVIRONMENT_DONE is not unset.

I don't think, at that time, this worked as you describe, but it was six years ago, and I haven't used zsh in years anyway, so my memory is not entirely trustworthy. If reverting my commit fixes mosh and tmux, and doesn't obviously break anything else, then go for it!
